include/lr.h

```diff
@@ -47,6 +47,25 @@ struct lr_cell {
                            // together in a circular fashion.
 };
 
+/* Provides a mechanism for a thread to exclusively access the linked ring. 
+ * The mutex should be recursive. 
+ * This means that the calling thread may call lock() multiple times before calling unlock(). 
+ * The mutex should be available to other threads only after the the thread that owns the mutex calls unlock() as many times as it called lock(). 
+*/
+struct lr_recursive_mutex
+{
+    /* Blocks until the mutex is acquired.
+     * Returns LR_OK on success and an error code otherwise. 
+     */
+    enum lr_result (*lock)();
```

Comment:
yes, i think having an argument would help. How about this?

```
struct mutex_attr
{
  void *state;
  enum lr_result (*lock)(void *state, lr_owner_t owner);
  enum lr_result (*unlock)(void *state);
};

lr_mutex_init(mutex_attr *attr);
```

In your example, we would have

```
struct mutex_state
{
    int *atomic_mutex_reg = 0xFFFF;
};

void mutex_lock(void *state, lr_owner_t owner)
{
    mutex_state *s = (mutex_state *) state;
    // Busy-wait until the mutex is not locked
    while ( *(s->atomic_mutex_reg) == 0 || *(s->atomic_mutex_reg) != owner) { }
    
    // Atomically set the mutex to locked
    *(s->atomic_mutex_reg) = owner;
}

void mutex_unlock(void *state)
{
    // Atomically set the mutex to unlocked
    mutex_state *s = (mutex_state *) state;
    *(s->atomic_mutex_reg) = owner;
}
```

Reply:
maybe `atomic_mutex_reg` should have the same type as `owner`, because bit wide is platform dependent. by default type is `uintptr_t`. In case of arbitrary bits wide, looks like using of default 0 value makes sense for clear and understandable initialization.

Reply:
thanks for the in-depth answers. I've updated the `lock` and `unlock` interfaces - please reopen or  add more comments if you see any issues.